WEB-95 Fix problems with GLIM application: saved total, "Moratorium", "Overdue Charges""

JaySoni1 opened the pull request:
**Changes Made :-** 

-Fixed GLIM loan application total calculation and ensured correct parent/child payload structure.
-Made Moratorium and Overdue Charges fields configurable and editable in the UI and API payload.

[WEB :- 95](https://mifosforge.jira.com/jira/software/c/projects/WEB/boards/62?assignee=712020%3A5685dc80-2da8-4d54-80e6-7b69c296926e&selectedIssue=WEB-95)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Edit overdue charge amounts directly in loan accounts.
  * Account creation now tracks application/group identifiers and first/last member handling.

* **Bug Fixes**
  * Safer rendering with guarded access to charge fields.
  * Consistent amount/currency display and improved charge update behavior.

* **Accessibility**
  * Added aria-labels to inputs and controls.
  * Standardized button types across steps and dialogs.

* **Refactor**
  * Simplified charge payloads and removed deprecated fields.

* **Chores**
  * Added i18n provider and updated login UI imports (FontAwesome + Material).
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

alberto-art3ch commented on src/app/login/login-form/login-form.component.ts:
```diff
@@ -10,7 +10,7 @@ import { AuthenticationService } from '../../core/authentication/authentication.
 import { FaIconComponent } from '@fortawesome/angular-fontawesome';
 import { MatIconButton, MatButton } from '@angular/material/button';
-import { MatCheckbox } from '@angular/material/checkbox';
+import { MatCheckboxModule } from '@angular/material/checkbox';
```
Why we are changing this login form in a Loan GLIM change?

JaySoni1 replied:
@alberto-art3ch  After making changes when I have started the server I found this error related to Angular Material imports comming when running the server to solve that I have made changes in this file .

alberto-art3ch reviewed (changes requested):
Please review my comment

steinwinde commented:
Hi @JaySoni1 , I'm very surprised that you were even able to test your changes - have you tested them? I downloaded your PR and tried to test it. However, when I chose "Applications" - "GLIM Application" from the hamburger menu, I got a Javascript error in the console:

> ERROR NullInjectorError: R3InjectorError(Standalone[_CreateGlimAccountComponent])[_I18nService -> _I18nService -> _I18nService -> _I18nService]: 
>   NullInjectorError: No provider for _I18nService!
>     at NullInjector.get (core.mjs:1600:21)
>     at R3Injector.get (core.mjs:2130:27)
>     at R3Injector.get (core.mjs:2130:27)
>     at R3Injector.get (core.mjs:2130:27)
>     at R3Injector.get (core.mjs:2130:27)
>     at ChainedInjector.get (core.mjs:16828:32)
>     at lookupTokenUsingModuleInjector (core.mjs:4955:31)
>     at getOrCreateInjectable (core.mjs:5001:10)
>     at Module.ɵɵdirectiveInject (core.mjs:16875:17)
>     at NodeInjectorFactory.CreateGlimAccountComponent_Factory [as factory] (create-glim-account.component.ts:44:40)
    
Only after making changes in i18n.service.ts, I was able to open the form for GLIM application. Have you not met this problem?

When I finally get to the "Active Client Members" tab/screen, I'm getting a series of Javascript errors

> ERROR TypeError: Cannot read properties of undefined (reading 'displaySymbol')
>     at LoansAccountPreviewStepComponent_div_216_td_14_Template (loans-account-preview-step.component.html:265:11)
>     at executeTemplate (core.mjs:12074:5)
>     at refreshView (core.mjs:13675:7)
>     at detectChangesInView (core.mjs:13887:5)
>     at detectChangesInViewIfAttached (core.mjs:13849:3)
>     at detectChangesInEmbeddedViews (core.mjs:13807:7)
>     at refreshView (core.mjs:13703:5)
>     at detectChangesInView (core.mjs:13887:5)
>     at detectChangesInViewIfAttached (core.mjs:13849:3)
>     at detectChangesInComponent (core.mjs:13837:3)

In the "loansAccount.charges" array, the charge seems to lack a currency, which causes the error, because the "displaySymbol" is retrieved from the currency. I'm not 100% sure this problem is related to your changes, but it looks to me as if it does. Because I'm not seeing this Javascript error when testing based on the code in the dev branch (but I see other Javascript errors). From what I can tell, it is the above Javascript error that causes the submit/next button to be disabled. I can't save a GLIM application, based on your PR.

steinwinde reviewed (changes requested):
Please don't use English text literals in the HTML ("Edit charge date" and many more). And please also see my other comment. I'm unable to test your PR. (We could have a video call and discuss your changes, if this helps.)

JaySoni1 commented:
Hi @steinwinde, Thank you for your feedback. I have now addressed the issues you mentioned. Yes, I did test the GLIM application flow earlier and did not encounter the erros at that time. However, after your review, I re-tested the code and was able to find the errors . I have now resolved the errors , please review .

steinwinde reviewed (changes requested):
@JaySoni1 , have you discussed this with @bharathcgowda  or @IOhacker  or someone competent? E.g. this adds two new, optional input fields on top of the Terms page - a position where we had the required "Principal" field before. Why do you think these two new input fields make sense? If you change significant parts of the application, please point out in the Jira ticket why. From my point of view these fields should rather appear under the Moratorium section (and their labels need translations!), but I don't even see why we need these fields.
In addition to that, please don't add English texts literally into the markup, e.g. `aria-label="Edit charge amount"`.

rhopman commented:
@JaySoni1 I did some digging into the GLIM application process on the backend. My understanding is as follows:
- You have to generate your own `applicationId` which serves to link the different loan applications into one GLIM.
- You have to make one request for each child loan (`m_loan` table). Each request should use the same `applicationId`.
- In the first request, you pass `isParentAccount: true` to create the parent account (`glim_accounts` table).
- In the last request, you pass `lastApplication: true` to close the GLIM (`accepting_child` will change to `0` in the `glim_accounts` table).
- Since we are using the `/batches` endpoint, we have to make sure that the requests are processed in order (or at least the first and last request should not be processed out of order). To do this, you can make each request dependent on the one before it, using the `reference` field. It is not entirely clear to me if this is absolutely necessary, but it seems like a safe choice.

All of this is reflected in the following code change in `create-glim-account.component.ts`:

```ts
  setData(applicationId: number, client: any, totalLoan: number, isFirst: boolean, isLast: boolean): any {
    const locale = this.settingsService.language.code;
    const dateFormat = this.settingsService.dateFormat;
    // const monthDayFormat = 'dd MMMM';
    const data = {
      ...this.loansAccount,
      charges: this.loansAccount.charges.map((charge: any) => ({
        chargeId: charge.id,
        amount: charge.amount
      })),
      clientId: client.id,
      totalLoan: totalLoan,
      loanType: 'glim',
      applicationId: applicationId,
      lastApplication: isLast,
      amortizationType: 1,
      principal: client.principal,
      syncDisbursementWithMeeting: false,
      expectedDisbursementDate: this.dateUtils.formatDate(this.loansAccount.expectedDisbursementDate, dateFormat),
      submittedOnDate: this.dateUtils.formatDate(this.loansAccount.submittedOnDate, dateFormat),
      dateFormat,
      // monthDayFormat,
      locale
    };
    data.groupId = this.loansAccountTemplate.group.id;

    if (isFirst) {
      // Note: Only add isParentAccount = true for the first request.
      // Setting isParentAccount = false for subsequent requests is not handled correctly by Fineract.
      data.isParentAccount = true;
    }

    delete data.principalAmount;
    // TODO: 2025-03-17: Apparently (?!) unsupported for GLIM
    delete data.allowPartialPeriodInterestCalculation;
    delete data.multiDisburseLoan;
    delete data.isFloatingInterestRate;

    return JSON.stringify(data);
  }

  /** Request Body Data */
  buildRequestData(): any[] {
    const requestData = [];
    const memberSelected = this.selectedMembers?.selectedMembers ?? [];
    const totalLoan = this.totalLoanAmount();

    // Create a random 10-digit applicationId for the GLIM
    const applicationId = Math.floor(1000000000 + Math.random() * 9000000000);

    for (let index = 0; index < memberSelected.length; index++) {
      const isFirst = index === 0;
      const isLast = index === memberSelected.length - 1;

      requestData.push({
        requestId: index,
        reference: index === 0 ? null : index - 1, // Make sure requests are executed in order
        method: 'POST',
        relativeUrl: 'loans',
        body: this.setData(applicationId, memberSelected[index], totalLoan, isFirst, isLast)
      });
    }
    return requestData;
  }
```

This only changes about a dozen lines from the `dev` branch. Of course, this does not fix the issues with "Moratorium" and "Overdue Charges", but at least the accounts are created in the correct way.

Please let me know if you have a different understanding of the process or API.

JaySoni1 commented:
@rhopman Thank you for your detailed summary and for reviewing the GLIM application process.
I have updated the PR . I have done the following things in the updated PR :- 

-I have generated a unique 10-digit applicationId and use it for all requests in the batch, linking all child loans to the same GLIM.
-Each child loan request uses the same applicationId.
-The first request includes isParentAccount: true to create the parent account and the last request includes lastApplication: true to close the GLIM.
-I have used the reference field in each request (except the first) to ensure the batch is processed in order, as you suggested.

If you have any  furthur suggestions please let me know .

JaySoni1 commented:
@steinwinde  Thank you for your feedback. I have not yet discussed these with @bharathcgowda or @IOhacke but I will reach out to them to confirm the necessity and placement of the new input fields I agree with your suggestion that these fields should be moved under the Moratorium section and that their labels should use translation keys. I will also update English texts along with the above changes .

IOhacker reviewed (approved):
LGTM, please resolve the conflict

JaySoni1 commented:
@IOhacker I will resolve the conflict and will update the PR soon

JaySoni1 committed:
WEB-95: WEB-95 Fix problems with GLIM application: saved total, "Moratorium", "Overdue Charges""

IOhacker commented on src/app/loans/loans-account-stepper/loans-account-charges-step/loans-account-charges-step.component.html:
```diff
@@ -24,7 +24,7 @@
       <th mat-header-cell *matHeaderCellDef>{{ 'labels.inputs.name' | translate }}</th>
       <td mat-cell *matCellDef="let charge">
-        {{ charge.name + ', ' + (charge.currency?.displaySymbol || '') }}
+        {{ charge.name }}<span *ngIf="charge.currency">, {{ charge.currency.displaySymbol }}</span>
```
have you done testing with loan accounts with penalties and charges?

IOhacker commented on src/app/loans/loans-account-stepper/loans-account-preview-step/loans-account-preview-step.component.html:
```diff
@@ -262,14 +262,14 @@ <h3 class="mat-h3 flex-fill">{{ 'labels.heading.Charges' | translate }}</h3>
         <th mat-header-cell *matHeaderCellDef>{{ 'labels.inputs.name' | translate }}</th>
         <td mat-cell *matCellDef="let charge">
-          {{ charge.name + ', ' + (charge.currency?.displaySymbol || '') }}
+          {{ charge.name }}<span *ngIf="charge.currency">, {{ charge.currency.displaySymbol }}</span>
```
have you done testing with loan products with penalties and charges?

IOhacker commented:
@JaySoni1 it is not working this is the error, please make sure it is being tested using the latest Apache Fineract: {
    "developerMessage": "The request was invalid. This typically will happen due to validation errors which are provided.",
    "httpStatusCode": "400",
    "defaultUserMessage": "Validation errors exist.",
    "userMessageGlobalisationCode": "validation.msg.validation.errors.exist",
    "errors": [
        {
            "defaultUserMessage": "The parameter moratoriumPrincipal is not supported.",
            "parameterName": "moratoriumPrincipal",
            "developerMessage": "The parameter moratoriumPrincipal is not supported.",
            "userMessageGlobalisationCode": "error.msg.parameter.unsupported",
            "args": [
                {
                    "value": "moratoriumPrincipal"
                }
            ]
        },
        {
            "defaultUserMessage": "The parameter moratoriumInterest is not supported.",
            "parameterName": "moratoriumInterest",
            "developerMessage": "The parameter moratoriumInterest is not supported.",
            "userMessageGlobalisationCode": "error.msg.parameter.unsupported",
            "args": [
                {
                    "value": "moratoriumInterest"
                }
            ]
        }
    ]
}

JaySoni1 commented:
Ok @IOhacker I will solve this and will update the PR soon.

Thank you for the review

IOhacker reviewed (approved):
LGTM

IOhacker commented:
@JaySoni1 prettier message

JaySoni1 commented:
Ok @IOhacker I will update the PR soon

JaySoni1 commented:
@IOhacker I have udated this PR please review

IOhacker reviewed (approved):
LGTM

IOhacker reviewed (approved):
LGTM

IOhacker commented:
@JaySoni1 there is one conflict here

IOhacker commented:
@JaySoni1 could you fix the conflic?

JaySoni1 commented:
@IOhacker I will fix this will update you soon

gkbishnoi07 commented:
@IOhacker can you please check this PR? are you merging this PR after conflict solve?

IOhacker commented:
@JaySoni1 please solve the conflcts, we will verify this PR.

JaySoni1 commented:
@IOhacker @gkbishnoi07 I have resolve the conflict and have updated the PR please review
